content/kb/troubleshooting-fstrim-error.md

```diff
@@ -0,0 +1,71 @@
+---
+title: "Troubleshooting: fstrim doesn't work on old kernel"
+author: Phan Le
+draft: false
+date: 2023-10-06
+catelogies:
+  - "trim"
+---
+
+## Applicable versions
+
+* Longhorn version >= `v1.4.0`
+* Node kernel version <= `v4.12`
+
+## Symptoms
+
+When running filesystem trim (either by Longhorn UI or manually by `fstrim` command on the host), it hits the error similar to:
+
+```
+unable to trim filesystem for volume pvc-e381424a-4866-447a-a75c-3096036f7846: cannot find volume pvc-e381424a-4866-447a-a75c-3096036f7846 mount info on host: failed to execute: nsenter [--mount=/host/proc/20357/ns/mnt --net=/host/proc/20357/ns/net fstrim /var/lib/kubelet/plugins/kubernetes.io/csi/driver.longhorn.io/d2df1133f3440486ddec39370380eeed7a3c71499981d63fc80e43c7ca9f4c9e/globalmount], output , stderr fstrim: /var/lib/kubelet/plugins/kubernetes.io/csi/driver.longhorn.io/d2df1133f3440486ddec39370380eeed7a3c71499981d63fc80e43c7ca9f4c9e/globalmount: FITRIM ioctl failed: Input/output error\n: exit status 255
+```
+or
+```
+unable to trim filesystem for volume pvc-e381424a-4866-447a-a75c-3096036f7846: cannot find volume pvc-e381424a-4866-447a-a75c-3096036f7846 mount info on host: failed to execute: nsenter [--mount=/host/proc/20357/ns/mnt --net=/host/proc/20357/ns/net fstrim /var/lib/kubelet/plugins/kubernetes.io/csi/driver.longhorn.io/d2df1133f3440486ddec39370380eeed7a3c71499981d63fc80e43c7ca9f4c9e/globalmount], output , stderr fstrim: /var/lib/kubelet/plugins/kubernetes.io/csi/driver.longhorn.io/d2df1133f3440486ddec39370380eeed7a3c71499981d63fc80e43c7ca9f4c9e/globalmount: the discard operation is not supported : exit status 1
+```
+
+## Details
+
+This is caused by a problem in old kernel SCSI driver which sets the wrong provisioning mode for the SCSI device that supports
```

Comment:
FYI
The legacy way in an old kernel is using write_same, but unfortunately, longhorn tgt doesn't implement the operation ([codes](https://github.com/rancher/tgt/blob/longhorn_oct2019/usr/bs_longhorn.c)). That's why the kernel falls back to `disabled`.

Reply:
Thanks for the link to the code!

Even though our engine doesn't support write_same, the created block device somehow still advertises that capability:
```
root@phan-v406-dt-pool2-e5bcaaa3-bxtvt:~# sg_vpd /dev/longhorn/testvol -p lbpv
Logical block provisioning VPD page (SBC):
  Unmap command supported (LBPU): 1
  Write same (16) with unmap bit supported (LBPWS): 1
  Write same (10) with unmap bit supported (LBPWS10): 1
  Logical block provisioning read zeros (LBPRZ): 1
  Anchored LBAs supported (ANC_SUP): 0
  Threshold exponent: 0 [threshold sets not supported]
  Descriptor present (DP): 0
  Minimum percentage: 0 [not reported]
  Provisioning type: 2 (thin provisioned)
  Threshold percentage: 0 [percentages not supported]
```

From old kernel perspective, it choose write_same over unmap which cause the problem. However, I am not sure how to modify LUN to disable advertising write_same capability. The only option I see it Thin-provisioning. When enable Thin-provisioning.

Reply:
I think it is due to that longhorn doesn't return `-1` to the upper caller for the unsupported write_same operation.
https://github.com/rancher/tgt/blob/longhorn_oct2019/usr/bs_longhorn.c#L199

Reply:
I am not sure if changing the return value over there can disable advertising the `write_same` capability because in the older Longhorn version (< v1.4.0), we don't return -1 but Longhorn volume still does NOT advertise the `write_same` capability. 

It is when we change `Thin-provisioning` yo `true` in Longhorn > v1.4.0, both unmap and write_same capabilitie are advertised  

Opening a ticket for further investigation https://github.com/longhorn/longhorn/issues/6876